keadm/cmd/keadm/app/cmd/edge/upgrade.go

```diff
@@ -0,0 +1,403 @@
+	"github.com/kubeedge/kubeedge/keadm/cmd/keadm/app/cmd/common"
+	"github.com/kubeedge/kubeedge/keadm/cmd/keadm/app/cmd/util"
+	"github.com/kubeedge/kubeedge/pkg/apis/componentconfig/edgecore/v1alpha1"
+	upgradev1alpha1 "github.com/kubeedge/kubeedge/pkg/apis/operations/v1alpha1"
+	"github.com/kubeedge/kubeedge/pkg/image"
+)
+
+var (
+	// idempotencyRecord is a file that is used to avoid upgrading node twice once a time.
+	// If the file exist, we don't allow upgrade node again
+	// we only allow upgrade nodes when the file NOT exist
+	idempotencyRecord = filepath.Join(util.KubeEdgePath, "idempotency_record")
+)
+
+// NewEdgeUpgrade returns KubeEdge edge upgrade command.
+func NewEdgeUpgrade() *cobra.Command {
+	upgradeOptions := newUpgradeOptions()
+
+	cmd := &cobra.Command{
+		Use:   "upgrade",
+		Short: "Upgrade edge component. Upgrade the edge node to the desired version.",
+		Long:  "Upgrade edge component. Upgrade the edge node to the desired version.",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			// upgrade edgecore
+			return upgradeOptions.upgrade()
+		},
+	}
+
+	addUpgradeFlags(cmd, upgradeOptions)
+	return cmd
+}
+
+// newJoinOptions returns a struct ready for being used for creating cmd join flags.
+func newUpgradeOptions() *UpgradeOptions {
+	opts := &UpgradeOptions{}
+	opts.ToVersion = "v" + common.DefaultKubeEdgeVersion
+	opts.Config = constants.DefaultConfigDir + "edgecore.yaml"
+
+	return opts
+}
+
+func (up *UpgradeOptions) upgrade() error {
+	// get EdgeCore configuration from edgecore.yaml config file
+	data, err := os.ReadFile(up.Config)
+	if err != nil {
+		return fmt.Errorf("failed to read config file %s: %v", up.Config, err)
+	}
+
+	configure := &v1alpha1.EdgeCoreConfig{}
+	err = yaml.Unmarshal(data, configure)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal config file %s: %v", up.Config, err)
+	}
+
+	upgrade := Upgrade{
+		UpgradeID:      up.UpgradeID,
+		HistoryID:      up.HistoryID,
+		FromVersion:    up.FromVersion,
+		ToVersion:      up.ToVersion,
+		Repository:     up.Repository,
+		ConfigFilePath: up.Config,
+		EdgeCoreConfig: configure,
+	}
+
+	defer func() {
+		// report upgrade result to cloudhub
+		if err := upgrade.reportUpgradeResult(); err != nil {
+			klog.Errorf("failed to report upgrade result to cloud: %v", err)
+		}
+		// cleanup idempotency record
+		if err := os.Remove(idempotencyRecord); err != nil {
+			klog.Errorf("failed to remove idempotency_record file(%s): %v", idempotencyRecord, err)
+		}
+	}()
+
+	// only allow upgrade when last upgrade finished
+	if util.FileExists(idempotencyRecord) {
+		upgrade.UpdateStatus(string(upgradev1alpha1.UpgradeFailedRollbackSuccess))
+		upgrade.UpdateFailureReason("last upgrade not finished, not allowed upgrade again")
+		return fmt.Errorf("last upgrade not finished, not allowed upgrade again")
+	}
+
+	// create idempotency_record file
+	os.MkdirAll(filepath.Dir(idempotencyRecord), 0755)
```

Comment:
1. Log the return error message for auditing.
2. File permission 0750 would be more secure?